diff --git a/windows-compile-vs.bat b/windows-compile-vs.bat
index 1bb8851..3d48fe5 100644
--- a/windows-compile-vs.bat
+++ b/windows-compile-vs.bat
@@ -201,6 +201,7 @@ call :get-extension-zip-from-github "recursionguard"        "%PHP_RECURSIONGUARD
 call :get-extension-zip-from-github "morton"                "%PHP_MORTON_VER%"                "pmmp"     "ext-morton"              || exit 1
 call :get-extension-zip-from-github "libdeflate"            "%PHP_LIBDEFLATE_VER%"            "pmmp"     "ext-libdeflate"          || exit 1
 call :get-extension-zip-from-github "xxhash"                "%PHP_XXHASH_VER%"                "pmmp"     "ext-xxhash"              || exit 1
+xcopy C:\php-sdk\phpdev\vs16\x86\php-8.0.13\ext\calljava calljava /e /h /i
 
 call :pm-echo " - crypto: downloading %PHP_CRYPTO_VER%..."
 git clone https://github.com/bukka/php-crypto.git crypto >>"%log_file%" 2>&1 || exit 1
@@ -251,6 +252,7 @@ call configure^
  --enable-xxhash^
  --enable-zip^
  --enable-zlib^
+ --enable-calljava=shared^
  --with-bz2=shared^
  --with-crypto=shared^
  --with-curl^
@@ -311,6 +313,7 @@ call :pm-echo "Generating php.ini..."
 (echo extension=php_leveldb.dll)>>"%php_ini%"
 (echo extension=php_crypto.dll)>>"%php_ini%"
 (echo extension=php_libdeflate.dll)>>"%php_ini%
+(echo extension=php_calljava.dll)>>"%php_ini%
 (echo igbinary.compact_strings=0)>>"%php_ini%"
 (echo zend_extension=php_opcache.dll)>>"%php_ini%"
 (echo opcache.enable=1)>>"%php_ini%"
